proposals/TP3-xds-error-propagation.md

```diff
@@ -0,0 +1,122 @@
+TP3: xds-error-propagation
+----
+* Author(s): anicr7
+* Approvers: htuch, markdroth, adisuissa
+* Implemented in: <xDS client, ...>
+* Last updated: 2024-10-25
+
+## Abstract
+
+This proposal introduces enhancements to the [xDS transport protocol](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol), focusing on improved debugging and observability. Specifically, it outlines a mechanism for the control plane to communicate error information directly to xDS clients.
+
+The objective of this proposal is to suggest a way for clients to receive feedback from xDS Management Servers in case of partial/drastic failures without closing any streams or connections.
+
+This proposal includes a new field for each subscribed Resource, called `ResourceError` which will provide detailed information for resource specific issues. The client must use this additional field to obtain notification for resources the xDS Management server couldn’t procure and provide necessary notification to the application. 
+
+## Background
+
+A frequent use case for xDS involves client subscription to multiple resources from the management server. Currently, if the xDS management server cannot provide a subset of these resources, the client experiences a complete loss of visibility. This can occur for various reasons, including but not limited to potential permission issues. This lack of granularity in the response can pose significant operational challenges for general client observability or debugging. 
```

Comment:
Suggested wording for this section:

"""
In cases where the xDS control plane is unable to send a resource that the client has subscribed to (e.g., because the resource does not exist or because the client lacks permission for the resource), the control plane does not have a good mechanism for conveying the cause of the error to the client.

The only existing mechanism for conveying a specific error message to a client is for the control plane to close the entire xDS stream with a non-OK status.  However, this is often undesirable behavior, because a problem with a single resource will also stop the client from receiving updates from other, unrelated resources.

Therefore, most control planes are forced simply act as if the resource does not exist.  In this situation, the client does not know why the control plane was unable to send the resource, so it cannot report useful information to human operators (e.g., via logs or request failure messages), which makes debugging challenging.

Note that while the incremental protocol variants can explicitly indicate to the client that a subscribed resource does not exist, the SotW protocol variants require the client to use a [15-second does-not-exist timeout](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#knowing-when-a-requested-resource-does-not-exist).  This adds an unnecessary delay before the client can react to the error.  It has also led to cases where control plane slowness has caused clients to incorrectly react as if a resource does not exist.
"""

Reply:
Done